### PR TITLE
plan: refine the range calculating.

### PR DESCRIPTION
**Pull request by hanfei1991**
For where a < 10 with index(a,b), the range will seek from (10, max), but will should let it seek from (10,min), Same with greater than.
@shenli @coocood @zimulala @tiancaiamao PTAL

### CODE REVIEW

File: plan/range.go [hunk 1]
```diff
@@ -317,7 +317,7 @@ func (r *rangeBuilder) newBuildFromPatternLike(expr *expression.ScalarFunction)
 			// Get the prefix, but exclude the prefix.
 			// e.g., "abc_x", the start point exclude "abc",
 			// because the string length is more than 3.
-			exclude = true
+			exclude = false
```
Comment (shenli) [inline plan/range.go]: Why change this to false?

File: plan/logical_plan_test.go [hunk 2]
```diff
@@ -822,7 +822,7 @@ func (s *testPlanSuite) TestRefine(c *C) {
 		},
 		{
 			sql:  "select a from t where c not between 1 and 2",
-			best: "Index(t.c_d_e)[[-inf,1) (2,+inf]]->Projection",
+			best: "Index(t.c_d_e)[[-inf <nil>,1 <nil>) (2 +inf,+inf +inf]]->Projection",
```
Comment (shenli) [inline plan/logical_plan_test.go]: Why there is nil?

Comment (hanfei1991) [inline plan/logical_plan_test.go]: c < 1 for index(c_d_e) shoule lead to range `[-inf <nil>,1 <nil>)` , for avoiding sweep the whole `1`

File: plan/statistics_helper.go [hunk 3]
```diff
@@ -24,18 +24,18 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
-func getRowCountByIndexRanges(sc *variable.StatementContext, statsTbl *statistics.Table, indexRanges []*IndexRange, indexInfo *model.IndexInfo, tableInfo *model.TableInfo) (uint64, error) {
+func (is *PhysicalIndexScan) getRowCountByIndexRanges(sc *variable.StatementContext, statsTbl *statistics.Table) (uint64, error) {
 	var offset int
-	for i := range tableInfo.Indices {
-		if tableInfo.Indices[i].Name.L == indexInfo.Name.L {
+	for i := range is.Table.Indices {
```
Comment (shenli) [inline plan/statistics_helper.go]: Only consider public indexes?

**Comment by shenli** [comment]
Rest LGTM
@coocood PTAL
@hanfei1991 Please make sure that tidb-test could pass.

File: plan/statistics_helper.go (hunk 3 above)
Comment (hanfei1991) [inline plan/statistics_helper.go]: Only public index will be the is.index

**Comment by shenli** [comment]
LGTM
@coocood @zimulala PTAL

File: plan/refiner.go [hunk 4]
```diff
@@ -64,6 +64,23 @@ func buildIndexRange(sc *variable.StatementContext, p *PhysicalIndexScan) error
 			refineRange(p.Ranges[i], p.Index)
 		}
 	}
+
+	if p.accessInAndEqCount+1 < len(p.Index.Columns) {
```
Comment (coocood) [inline plan/refiner.go]: Can we just check if the length of high value or low value of each range has the same length as index columns, then append nil or max value?

Comment (hanfei1991) [inline plan/refiner.go]: anyway is ok

Comment (coocood) [inline plan/refiner.go]: It is easier to understand.

File: plan/refiner.go [hunk 5]
```diff
@@ -65,7 +65,7 @@ func buildIndexRange(sc *variable.StatementContext, p *PhysicalIndexScan) error
 		}
 	}
 
-	if p.accessInAndEqCount+1 < len(p.Index.Columns) {
+	if len(p.Ranges[0].LowVal) < len(p.Index.Columns) {
 		for _, ran := range p.Ranges {
 			if ran.HighExclude || ran.LowExclude {
```
Comment (coocood) [inline plan/refiner.go]: Why only process Exclude value?
Non-exclude value need to append nil or max too?
Can we make sure every range value have the same length as index columns?

Comment (hanfei1991) [inline plan/refiner.go]: 1. Non-exclude needn't to change
2. we cannot make sure that, may be still less than the len of index cols.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@hanfei1991 
What I mean is that if we enforce every index range to be the same length as index columns, then we will solve this problem as a side effect, but that is much easier to understand.

And with the guarantee, the index range will be much easier to use later.

**Comment by hanfei1991** [comment]
We can consider to do that in future :) @coocood 
Please approve this PR.

**Comment by ngaut** [comment]
Could you file an issue to track it ?  @hanfei1991

**Comment by hanfei1991** [comment]
#2541

File: executor/executor_distsql.go [hunk 6]
```diff
@@ -218,7 +217,7 @@ func convertIndexRangeTypes(sc *variable.StatementContext, ran *plan.IndexRange,
 		break
 	}
 	for i := range ran.HighVal {
-		if ran.HighVal[i].Kind() == types.KindMaxValue {
+		if ran.HighVal[i].Kind() == types.KindMaxValue || ran.LowVal[i].Kind() == types.KindNull {
```
Comment (jackysp) [inline executor/executor_distsql.go]: Why skip LowVal Kind == KindNull in HighVal range loop?

**Pull request merged**